lcd.lua

```diff
@@ -171,6 +171,56 @@ local generate_texture = function(lines)
 	return texture
 end
 
+local number_to_color = function(num)
+	if type(num) ~= "number" or num >= 4096 or num < 0 then
+		return "#0000"
+	end
+
+	return string.format("#%X%X%X", num / 16^2, num / 16 % 16, num % 16 )
+end
+
+local generate_raster_texture = function(colors)
+	-- the LCD texture is 16x16, will upscale the resolution by 4
+	local total_width = 16 * 4
+	local w_padding = 1 * 4
+	local h_padding = 2 * 4
+
+	local width = total_width - w_padding * 2
+	local height = total_width - h_padding * 2
+
+	local texture = "[fill:"..total_width.."x"..total_width..":#0000"
+
+	-- the LCD's frame
+	-- top
+	texture = texture
+		.."^[fill:"..total_width.."x"..h_padding
+		..":0,0:#000"
+	-- bottom
+	texture = texture
+		.."^[fill:"..total_width.."x"..h_padding
+		..":0,"..total_width - h_padding..":#000"
+	-- left
+	texture = texture
+		.."^[fill:"..w_padding.."x"..total_width - h_padding
+		..":0,"..h_padding..":#000"
+	-- right
+	texture = texture
+		.."^[fill:"..w_padding.."x"..total_width - h_padding
+		..":".. total_width - w_padding ..","..h_padding..":#000"
+
+	if type(colors) == "table" then
+		for i, num in ipairs(colors) do
+			if i > width * height then break end
+			local color = number_to_color(num)
+			local x = (i - 1) % width
+			local y = math.floor((i - 1) / width)
+			texture = texture .. "^[fill:1x1:"..x + w_padding ..","..y + h_padding..":"..color
```

Comment:
Please use a table and `table.concat` the thing to a string at the end. This kind of string concat scales badly the longer the string gets.